Runnable Examples: 21 total, added 10

rasmus-kirk opened the pull request:
## Purpose

See #153 

## Changes

Added following runnable examples:

- `getBlockFinalizationSummary.ts`
- `getBlockInfo.ts`
- `getBlockPendingUpdates.ts`
- `getBlockSpecialEvents.ts`
- `getBlockTransactionEvents.ts`
- `getBlocks.ts`
- `getBlocksAtHeightAbsolute.ts`
- `getBlocksAtHeightRelative.ts`
- `getBranches.ts`
- `getConsensusStatus.ts`

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

abizjak commented on examples/client/getBlockFinalizationSummary.ts:
```diff
@@ -0,0 +1,59 @@
+/// If a blockhash is not supplied it will pick the latest finalized block.
+
+(async () => {
+    const pendingUpdates: AsyncIterable<PendingUpdate> =
```
The name of this file does not fit what is being done here.

abizjak commented on examples/client/getBlockInfo.ts:
```diff
@@ -0,0 +1,56 @@
+
+    console.dir(blockInfo, { depth: null, colors: true });
+
+    // The blockInfo contain information that can then be extracted:
```
We could do a bit more effort, such as printing "a block with N transactions baked at ?. It's parent is ...". Or some such.

rasmus-kirk replied:
I could do a few of those. Do we then want to keep line 52 that prints the whole block then?

The direction I've been going so far is to print the entire returned object, and if applicable, extract information as unused `const`'s that are not printed. Do we want to print things like this or do I continue defining them as unused `const`'s but with better documentation/comments?

abizjak replied:
It's up to you, but I would go for something like printing a few interesting things only. Dumping large objects is not super nice. They should get intellisense support in VSCode I assume when modifying the example.

abizjak commented on examples/client/getBlocks.ts:
```diff
@@ -0,0 +1,54 @@
+        console.dir(block, { depth: null, colors: true });
+    }
+
+    // Can also be collected to a list with:
+    const blockList: ArrivedBlockInfo[] = await streamToList(blocks);
```
I'm not sure this will work seeing that the stream never ends.

abizjak commented on examples/client/getBlocksAtHeightRelative.ts:
```diff
@@ -0,0 +1,72 @@
+
+(async () => {
+    const request: BlocksAtHeightRequest = {
+        genesisIndex: 1,
```
This should not be hardcoded. It should also be explained.

rasmus-kirk replied:
Of course should be `cli.flags.genesisIndex` I'll also flesh out its description on line 16 further. Do you want some explanation here as well.

abizjak replied:
yes. Always :)

abizjak commented on examples/client/getBlocksAtHeightRelative.ts:
```diff
@@ -0,0 +1,72 @@
+    const request: BlocksAtHeightRequest = {
+        genesisIndex: 1,
+        height: BigInt(cli.flags.height),
+        restrict: cli.flags.restrict,
```
Also explain.

rasmus-kirk replied:
Is the explanation on line 14-15 not enough or do you also want a description here.

soerenbf commented on examples/client/getBlockFinalizationSummary.ts:
```diff
@@ -0,0 +1,64 @@
+    cli.showHelp();
+}
+
+/// Get the summary of the finalization data in a given block. Only finalized
+/// blocks will return a finalization summary, if the summary is requested for
+/// a non-finalized block, this will return an object with only the tag field,
+/// with value "none".  If a blockhash is not supplied it will pick the latest
+/// finalized block.
```
Triple slash comments are not docstrings in typescript (I assume you have this from rust?), but is actually a format reserved for type references. I think we should use proper JS block comments instead

```ts
/**
 * Block comment goes here
 */
```

Same goes for all other files in the PR

rasmus-kirk replied:
I knew triple slashes was not "supported" but I still wanted to differentiate between documentation and comments, I did not know triple slashes were reserved however. Using the JS blocks are better.

I'll `sed` it in #164. I'll change all imports to only be `node-sdk` for these client examples in that same PR.

rasmus-kirk replied:
Done now

soerenbf reviewed (changes requested):
I think we should only have references to either `@concordium/node-sdk` or `@concordium/web-sdk` on a per file basis, as developers using the dependency will only have `@concordium/common-sdk` implicitly installed (as a dependency of a dependency)

shjortConcordium reviewed (commented):
I agree with @soerenbf 's comment, it makes sense to not use common-sdk directly, because the users won't (/shouldn't) do that.

otherwise LGTM

soerenbf commented on examples/client/getBlocks.ts:
```diff
@@ -0,0 +1,51 @@
+    address,
+    Number(port),
+    credentials.createInsecure(),
+    { timeout: 15000 }
```
Yeah.. people will ctrl-C anyhow if it runs for too long 😄

soerenbf reviewed (approved):
LGTM by now and with the promise of including implementations of changes in resolved comments in subsequent PR's.

I think, however, we should limit this workflow (fixing requested changes for one PR in another one) to the absolute minimum, as it in essence circumvents the purpose of approving changes to be included in the target branch, when in fact the changes are merely promised to be implemented at a later point in time.

If you want someone else to also approve, wait with merging until they've had a chance to look over the changes as well.

abizjak commented:
I support @soerenbf in expressing that it is almost never OK to say "I'll fix this in a followup PR". Structure PRs in a way that that is not necessary.

abizjak commented on examples/.eslintrc.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "rules": {
+    "@typescript-eslint/no-unused-vars": 0
```
I don't think you should do this. Print them instead./

abizjak reviewed (approved):
Address the existing comments. Otherwise looks ok.

rasmus-kirk commented:
All right I addressed the comments and some other issues I stumbled on: 
- Printed more information wherever I saw it fit
- Removed the `-h` flag, it had inconsistent behavior with `--help`
- Removed all unused `const`'s
- Don't do `streamToList` on stream examples, but mention it in the documentation.
- I don't import from `common` anymore, only `nodejs`
- Documentation comments now have correct style
